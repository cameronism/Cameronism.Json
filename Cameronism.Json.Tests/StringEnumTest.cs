using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Cameronism.Json.Tests
{
	public unsafe class StringEnumTest
	{
		[Fact]
		public void DayOfWeekEnum()
		{
			var buffer = new byte[1024];
			int stringStart;
			int freeStart;

			fixed (byte* ptr = buffer)
			{
				var lookup = StringEnum.GenerateSequentialLookup(typeof(DayOfWeek), ptr, buffer.Length, out freeStart);

				Assert.True(lookup.HasValue, "lookup must have been generated");
				stringStart = (int)(lookup.Value.StringStart - ptr);
				
				//var chars = new char[64];
				//foreach (DayOfWeek dow in Enum.GetValues(typeof(DayOfWeek)))
				//{
				//	byte* str;
				//	int len;
				//	lookup.Value.FindSequential((ulong)dow, out str, out len);
				//	Assert.True(str != null, "string must not be null");
				//	var dowString = dow.ToString();
				//	Assert.Equal(dowString.Length, len);

				//	string actualString;
				//	fixed (char* ch = chars)
				//	{
				//		var strLen = Encoding.UTF8.GetChars(str, len, ch, chars.Length);
				//		actualString = new string(ch, 0, strLen);
				//	}
				//	Assert.Equal(dowString, actualString);
				//}
			}

			var sb = new StringBuilder();
			sb.AppendLine("## Lookups");
			Util.Hex.Dump(sb, buffer.Take(stringStart));
			sb.AppendLine();
			sb.AppendLine("## Strings");
			Util.Hex.Dump(sb, buffer.Skip(stringStart).Take(freeStart - stringStart));

			ApprovalTests.Approvals.Verify(sb.ToString());
		}
	}
}
