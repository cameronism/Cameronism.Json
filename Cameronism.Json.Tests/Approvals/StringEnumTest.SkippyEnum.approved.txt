# Indexed Cameronism.Json.Tests.StringEnumTest+Skippy
## Lookups


0000   00 00 00 00 00 00 03 00   00 00 00 00 03 00 03 00   ........ ........
0010   00 00 00 00 06 00 03 00   00 00 00 00 09 00 03 00   ........ ........

## Strings


0000   46 6f 6f 42 61 72 42 6f   70 42 61 7a               FooBarBo pBaz

# Sorted Cameronism.Json.Tests.StringEnumTest+Skippy
## Lookups


0000   01 00 00 00 00 00 03 00   03 00 00 00 03 00 03 00   ........ ........
0010   05 00 00 00 06 00 03 00   07 00 00 00 09 00 03 00   ........ ........

## Strings


0000   46 6f 6f 42 61 72 42 6f   70 42 61 7a               FooBarBo pBaz